lib/datadog/tracing/contrib/http/distributed/propagation.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+# typed: true
+
+require_relative '../../../distributed/propagation'
+require_relative 'b3'
+require_relative 'b3_single'
+require_relative 'datadog'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module HTTP
+        module Distributed
+          # Extracts and injects propagation through HTTP headers.
+          class Propagation < Tracing::Distributed::Propagation
+            def initialize
```

Comment:
I didn't have time to do the whole thing, which involves touching `components.rb` and managing the lifecycle of `Propagation`.

The changes that are needed in the future, to perform that correct dependency injection, will work directly with these classes in this PR: they are now classes, not singletons, and are in their correct namespace.

Reply:
I'll admit that I'm not convinced these "midway points" are useful, but I'll trust you this one :)

Reply:
> I'll admit that I'm not convinced these "midway points" are useful, but I'll trust you this one :)

I had to break down the work somehow, given I already put more time that I budgeted for for this refactor, so I chose a stopping point that allowed me move forward with adding new propagators without pulling hairs.
In an ideal scenario, the whole thing would be refactored, but I'd probably need another week to do it.

Reply:
Mega fair! Hope you didn't take the above as criticism of you, it's more of a "I suspect this detour may need to be slightly redone and so I would prefer omitting it, but I don't have any data/proof to match my suspicion and this is reasonable as well so since we need to make a decision, I trust yours."